docs/madr/decisions/046-meshservice-multizone.md

```diff
@@ -0,0 +1,361 @@
+      app: redis
+      k8s.kuma.io/namespace: redis-system
+      kuma.io/zone: east-1
+  ports: ...
+status: ...
+```
+
+This will be then synced to other zones in this form, which is the same syncing strategy that we already have with ZoneIngress.
+
+Because of the hashes and syncing only to system namespace, we avoid clashes of synced services and existing services in the cluster.
+
+**Hostname and VIP management**
+
+Advertising the service is only part of the solution. We also need to provide a way to consume this service.
+
+We cannot reuse:
+* existing hostnames like `redis.redis-system.svc.cluster.local`, because there might be already a redis in cluster `east.
+* ClusterIP, because each cluster network is isolated. We might clash with IPs.
+
+Therefore `status` object should not be synced from zone `east` to zone `west`.
+
+To provide hostname, we can use `HostnameGenerator` described in MADR #XXX (todo provide a number when merged).
+```yaml
+type: HostnameGenerator
+name: k8s-zone-hostnames
+spec:
+  template: {{ label "kuma.io/display-name" }}.{{ label "k8s.kuma.io/namespace" }}.svc.mesh.{{ label "kuma.io/zone" }}
+```
+so that the synced service in zone `west` gets this hostname `redis.redis-system.svc.mesh.east`.
+
+To assign Kuma VIP, we follow the process descirbed in MADR #XXX (todo provide a number when merged).
+
+When generating the XDS configuration for the client in zone `west`, we check if MeshService has `kuma.io/zone` tag that is not equal to local zone.
+If so, then we need to point Envoy Cluster to endpoint for ZoneIngress of `west`.
+This way we get rid of `ZoneIngress#availableServices`, which was one of the initial motivation of introducing MeshService (see MADR #YYY).
+
+#### Service export / import
+
+One thing to consider is whether we should sync all MeshServices to other clusters by default.
+It's reasonable to consider that only a subset of services is consumed cross zone.
+
+Trimming it could benefit in:
+* performance. Fewer objects to store and sync around.
+  However, changes to MeshServices spec are rather rare. We don't consider this to be a significant perf issue.
+* security. Unexported Mesh Service would not even be exposed with ZoneIngress.
+
+Explicit import seems to have very little value, because clients needs to explicitly use service from a specific zone.
+
+##### Export field
+
+We can introduce `spec.export.mode` field to MeshService with values `All`, `None`.
+We want to have nested mode in export, so we can later extend api if needed. For example `spec.export.mode: Subset` and `spec.export.list: zone-a,zone-b`.
+On Kubernetes, this would be controlled via `kuma.io/export-mode` annotation on `Service` object.
+On Universal, where MeshService is synthesized from Dataplane objects, this would be controlled via `kuma.io/export-mode` tag on a Dataplane object.
+
+If missing, then default behavior from kuma-cp configuration will be applied. The default CP configuration is `spec.export.mode: All`.
```

Comment:
```suggestion
If missing, then default behavior from kuma-cp configuration will be applied. The default zone CP configuration is `spec.export.mode: All`.
```
Actually maybe don't use a CP configuration for this just because it creates complexity with switching the default etc no?

We could be opinionated what the default is and folks that want can set it on each policy differently. It's both simpler and more straightforward (plus that's how k8s work)

Reply:
We can do this. We can add the setting if needed later.

Reply:
Admission controllers/Kyverno and friends are here for this. I'm a little worried of folks suddenly switching this and appreciate having less moving parts

Reply:
> suddenly switching this

how?

> appreciate having less moving parts

maybe. Although default defined in CP config is something we do very often

Reply:
ok if you think so.